src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/DiagnosticCounter.cs

```diff
@@ -125,12 +125,32 @@ internal string GetMetadataString()
                 return "";
             }
 
-            StringBuilder sb = new StringBuilder("");
-            foreach(KeyValuePair<string, string> kvPair in _metadata)
+            // The dictionary is only initialized to non-null when there's metadata to add, and no items
+            // are ever removed, so if the dictionary is non-null, there must also be at least one element.
+            Dictionary<string, string>.Enumerator enumerator = _metadata.GetEnumerator();
+            Debug.Assert(_metadata.Count > 0);
+            bool gotOne = enumerator.MoveNext();
+            Debug.Assert(gotOne);
+
+            // If there's only one element, just concat a string for it.
+            KeyValuePair<string, string> current = enumerator.Current;
+            if (!enumerator.MoveNext())
             {
-                sb.Append($"{kvPair.Key}:{kvPair.Value},");
+                return current.Key + ":" + current.Value;
             }
-            return sb.Length == 0 ? "" : sb.ToString(0, sb.Length - 1); // Get rid of the last ","
+
+            // Otherwise, append it, then append the element we moved to, and then
+            // iterate through the remainder of the elements, appending each.
+            var sb = new StringBuilder().Append(current.Key).Append(':').Append(current.Value);
```

Comment:
cached string builder maybe?

Reply:
From PR description:
"If they're really important, we could subsequently take it a step further and store the StringBuilder instance onto the DiagnosticCounter instance, or use a ValueStringBuilder, or so on.)"

Reply:
I mean, is there a reason to not use StringBuilderCache if the string  will normally fit? But, no matter.

Reply:
If it doesn't fit, it'll ruin it for everyone else.